model/riscv_vmem_sv48.sail

```diff
@@ -1,5 +1,18 @@
 /* Sv48 address translation for RV64. */
 
+val napot_bits48 : (bits(44), nat) -> nat
+function napot_bits48 (v, n) = {
+  let mask = shiftl(v ^ v ^ EXTZ(0b1), n + 1) - 1;
+  let pattern = shiftl(v ^ v ^ EXTZ(0b1), n);
+  if n > 8 then {
+    0
+  } else if (v & mask) == pattern then {
+    n + 1
+  } else {
+    napot_bits39(v, n + 1)
```

Comment:
Wrong function? And how is the implementation of this meant to differ from the Sv48 version? Can they share a common implementation?

Reply:
Oops, yes, this should be napot_bits48.  Good catch.

In general almost all of the code in Sv32, Sv39, and Sv48 is redundant already.  I agree it would make sense to factor out not just this new napot_bits function but also all of the rest of that common code, but I didn't want to tie that into this Zsn PR or take all of that on at one in my first attempt to write Sail, so I kept it separate for now.

Reply:
Yeah, it's just slightly more awkward with the extra level of translation (but easy enough to do), and Sv32 is rather different due to not having the upper bits in the PTEs. Merging Sv39 and Sv48 together is easy enough to do, but for Sv32 it'd probably be a case of pulling out common functionality but still having separate files, one for Sv32 and one for Sv39+48(+57).

Reply:
I'm willing to give this refactoring a shot (on a separate standalone commit).  We're also planning to add Sv57 soon, so it would make sense to refactor this now before duplicating the code yet again.

I do have one question though @jrtc27.  Within the walkN functions, while the code is mostly the same, there are function calls like Mk_SV48_PTE(v) which are specific to the PTE type, and the Sv32, Sv39, Sv48, and Sv57 (assuming I follow the pattern) PTEs are defined as different types in riscv_vmem_common.sail, even though all but Sv32 are actually the same.  If I make the function generic, I don't see an easy way to call the right Mk_SVN_PTE function for the given PTE type.  I can't just factor it out from the top either because the function is called recursively.

Is there an easy way in Sail to write a function like walk in a generic way while still being able to specialize for the right Mk_SVN_PTE(v) function internally?  Similar question elsewhere, e.g., for translate48 and add_to_TLB48.  I couldn't figure out how to pass a function pointer as an argument, if that's even possible in Sail.  Are there other ways around this?  If not, perhaps that's why the code was just duplicated in the first place?

Alternatively, would it be a bad thing to at least simply merge the Sv39, Sv48, and Sv57 PTE types into common "64b PTE" types, so at least we'd get this down to 2x duplication (32b vs. 64b) instead of 4x or more?

Reply:
Hello Daniel, what's the state of the refactoring you mentioned?

Reply:
Hi @martinberger, I haven't had a chance to look at this since my last comment, in part because I didn't (and still don't) have enough Sail expertise to figure out how to address the questions I asked in my last comment.  It still seems like it would be a good idea to refactor the code to remove redundancy between Sv39, Sv48, and (imminently) Sv57 first, but I'm not really sure how to do it.

Also, I should point out that "Zsn" is now called "Svnapot"...the functionality hasn't changed; only the name has changed.

Reply:
@daniellustig How imminent is Sv57?

Maybe it makes sense to  close the PR and resubmit

- if/when Sv57 is added, and
- either the redundancies have been removed, or removing them is not possible?

Note that (for good reasons) Sail has only limited abstraction facilities, so the redundancies might be needed.

Reply:
Closing the PR is fine with me if that's what you prefer.  Is someone going to look into either removing the redundancies or declaring that we have to live with the redundancy?

FWIW Sv57 is currently out for public review, so it will very likely be ratified soon, even if it's not yet in the Sail model.  https://groups.google.com/a/groups.riscv.org/g/isa-dev/c/1bTAL-PJGYQ

Reply:
Actually, all Priv instructions like Sv57 have a standing waiver for this year for SAIL work.  But, the advice still stands: do what's right for the project.

Reply:
@daniellustig @jjscheel I don't think it's much work, maybe 100 lines of Sail, so it would be easy to do for somebody who was well-versed in VM and Sail ... Anyway, it sounds like we should not wait until Sv57 is ready and keep it separate.

Reply:
@daniellustig Did you have further thoughts on this? I'm happy to support  Sv57 getting in as soon as possible.

Reply:
@daniellustig I have no strong preference. It would be good to speed up merging important PRs in.

Reply:
@daniellustig I've had some undergraduate summer interns who did a lot of Sail. They are back at university, I can ask them if they wanted to do (paid) Sail work, if you like.

Reply:
Hi @martinberger, sorry for the delayed response (it was the Thanksgiving holiday here in the US).  If they are interested and funding is available, then it would be great to have their help.

Reply:
@daniellustig I don't know about funding, I doubt that the RISCV organisation pays this. But he is interested, I'll connect you in a PM.

Reply:
Mark Himelstein is the person to talk to about that. I know there is money floating around for some things like this, but don't recall exactly what.